File: LAB7/python/main.py
import cv2
import time
  
  
# define a video capture object
vid = cv2.VideoCapture(0)
  
while(True):
      
    # Capture the video frame
    # by frame
    ret, frame = vid.read()
    start = time.time()
    # Flip the video
    frame = cv2.flip(frame, 1)
    

    img_gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
    blur_img = cv2.GaussianBlur(img_gray, (3,3) ,0)
    ret, img_thresh = cv2.threshold(blur_img, 150, 255, cv2.THRESH_BINARY)
    # detect contours on binary image
    contours, _ = cv2.findContours(img_thresh, cv2.RETR_TREE, cv2.CHAIN_APPROX_NONE)

    cv2.drawContours(frame, contours, -1, (0, 255, 0), 2)
    if len(contours)>0:
        # find the biggest countour (c) by the area
        c = max(contours, key = cv2.contourArea)
        end = time.time()
        x,y,w,h = cv2.boundingRect(c)

        # draw the biggest contour (c) in green
        cv2.rectangle(frame,(x,y),(x+w,y+h),(255,0,0),2)
    cv2.imshow('frame',frame)
    #Exit on ESC
    if cv2.waitKey(1) & 0xFF == 27:
        print("Time: ",end - start)
        break
  
# After the loop release the cap object
vid.release()
# Destroy all the windows
cv2.destroyAllWindows()

File: LAB7/cpp/main.cpp
#include <opencv2/opencv.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/core.hpp>
#include <iostream>
#include <vector>
#include <chrono>
// g++ pkg-config --cflags opencv4 cv2.cpp pkg-config --libs opencv4
using namespace std;
using namespace cv;

int main(int argc, char **argv)
{
    using Clock = std::chrono::steady_clock;
    std::chrono::time_point<std::chrono::steady_clock> start;
    std::chrono::time_point<std::chrono::steady_clock> now;
    std::chrono::milliseconds duration;

    cv::Mat input_grey, blur_img, thresh_img;
    cv::Size kernel(3, 3);

    VideoCapture vc(1);
    vc.open(1, CAP_DSHOW);
    if (!vc.isOpened())
        return 1;
    cv::Mat frame;

    int i = 0;
    while (waitKey(10) != 27)
    {
        vc >> frame;
        start = Clock::now();
        flip(frame, frame, 1);

        cv::cvtColor(frame, input_grey, cv::COLOR_BGR2GRAY);
        cv::GaussianBlur(input_grey, blur_img, kernel, 0);
        cv::threshold(blur_img, thresh_img, 150, 255, cv::THRESH_BINARY);

        std::vector<std::vector<cv::Point>> contours;
        std::vector<cv::Vec4i> hierarchy;

        cv::findContours(thresh_img, contours, hierarchy, cv::RETR_TREE, cv::CHAIN_APPROX_SIMPLE);

        cv::Scalar green(0, 255, 0);
        cv::drawContours(frame, contours, -1, green, 2);

        int largest_area = 0;
        int largest_contour_index = 0;
        Rect bounding_rect;
        for (size_t i = 0; i < contours.size(); i++) // iterate through each contour.
        {
            double area = contourArea(contours[i]); //  Find the area of contour

            if (area > largest_area)
            {
                largest_area = area;
                largest_contour_index = i;                 // Store the index of largest contour
                bounding_rect = boundingRect(contours[i]); // Find the bounding rectangle for biggest contour
            }
        }
        now = Clock::now();
        cv::rectangle(frame, bounding_rect, Scalar(255, 0, 0), 2);

        cv::imshow("Contours", frame);
    }
    duration = std::chrono::duration_cast<std::chrono::milliseconds>(now - start);
    std::cout << "Time: " << duration.count() << " ms." << endl;

    vc.release();
    cv::destroyAllWindows();
    return 0;
}